src/main/java/org/broadinstitute/hellbender/utils/fragments/FragmentCollection.java

```diff
@@ -0,0 +1,144 @@
+     * The callers must not modify those arguments after handing them off to this collection.
+     *
+     * The constructor is private - use the factory method if you need an object.
+     */
+    private FragmentCollection(final Collection<T> singletons, final Collection<List<T>> overlappingPairs) {
+        this.singletons = singletons == null ? Collections.emptyList() : singletons;
+        this.overlappingPairs = overlappingPairs == null ? Collections.emptyList() : overlappingPairs;
+    }
+
+    /**
+     * Gets the T elements not containing overlapping elements, in no particular order.
+     * The returned collection is unmodifiable.
+     */
+    public Collection<T> getSingletonReads() {
+        return Collections.unmodifiableCollection(singletons);
+    }
+
+    /**
+     * Gets the T elements containing overlapping elements, in no particular order
+     * The returned collection is unmodifiable.
+     */
+    public Collection<List<T>> getOverlappingPairs() {
+        return Collections.unmodifiableCollection(overlappingPairs);
+    }
+
+    /**
+     * Generic algorithm that takes an iterable over T objects, a getter routine to extract the reads in T,
+     * and returns a FragmentCollection that contains the T objects whose underlying reads either overlap (or
+     * not) with their mate pairs.
+     *
+     * @param readContainingObjects An iterator of objects that contain SAMRecords
+     * @param nElements the number of elements to be provided by the iterator, which is usually known upfront and
+     *                  greatly improves the efficiency of the fragment calculation
+     * @param getter a helper function that takes an object of type T and returns is associated SAMRecord
+     * @param <T>
+     * @return a fragment collection
+     */
+    private static <T> FragmentCollection<T> create(final Iterable<T> readContainingObjects, final int nElements, final Function<T, GATKRead> getter) {
+        Collection<T> singletons = null;
+        Collection<List<T>> overlapping = null;
+        Map<String, T> nameMap = null;
+
+        int lastStart = -1;
+
+        // build an initial map, grabbing all of the multi-read fragments
+        for ( final T p : readContainingObjects ) {
+            final GATKRead read = getter.apply(p);
+
+            if ( read.getStart() < lastStart ) {
+                throw new IllegalArgumentException(String.format(
+                        "FragmentUtils.create assumes that the incoming objects are ordered by " +
+                                "SAMRecord alignment start, but saw a read %s with alignment start " +
+                                "%d before the previous start %d", read.getName(), read.getStart(), lastStart));
+            }
+            lastStart = read.getStart();
+
+
+            if ( ! read.isPaired() || read.mateIsUnmapped() || read.getMateStart() == 0 || read.getMateStart() > read.getEnd() ) {
```

Comment:
`read.getMateStart() == 0` implies `read.mateIsUnmapped()` in the `GATKRead` interface, so this is redundant.